### PR TITLE
Wire timeline search-result highlighting across cards, modal, and tool details

### PR DESCRIPTION
**Pull request by the-non-expert**
## Summary
- Searching a session's timeline found matches but never highlighted them once you opened a result — this wires the search query into every render path that can show one: collapsed card, inline-expanded card, modal popup, `ToolCallDetail`'s per-tool panels, and todo updates.
- Broadens which fields are actually searchable/renderable (Edit `old_string`/`new_string`, Task/Agent `prompt`), with server-side truncation so those additions don't balloon the 1s live-session poll payload.
- Fixes the correctness, performance, and security issues surfaced by two rounds of independent blind code review, including:
  - A per-event search cache that never hit on live sessions (keyed on object identity, which polling replaces every second).
  - A highlight/escape ordering bug where certain queries (`<div>`, `&&`, `"key"`) never highlighted, and others (`lt`, `amp`) corrupted rendered text.
  - A debounce race that could resurrect a search query right after it was cleared via Escape.
  - Untruncated diffs/prompts making highlighting run over multi-MB strings that are clipped to a 320px box anyway.
  - Match-count vs. what's-actually-highlighted divergence, a stuck "hidden match" indicator, whitespace-only-query false positives, and a pre-existing `input.file_path`/`path` key mismatch that left Read/Write/Edit paths blank.

## Test plan
- [x] `npm run check` (frontend) — 0 errors, same 13 pre-existing warnings
- [x] `pytest` (api) — 1795 passed, same 2 pre-existing unrelated failures (local `.storygraph.db` leak, not caused by this branch)
- [x] Both dev servers verified healthy after changes
- [x] Manual verification in-browser (search across collapsed/expanded/modal/tool-detail views, live session behavior)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: api/utils.py [hunk 1]
```diff
@@ -749,6 +749,18 @@ def find_best_root(path: str, working_dirs: list[str]) -> str | None:
         return None
 
 
+# Cap on free-text tool-input fields (diffs, subagent prompts) placed into
+# timeline metadata. Without this, large Edit diffs or Task prompts ship in
+# full over the timeline endpoint, which live sessions poll every second.
+_METADATA_TEXT_LIMIT = 4000
+
+
+def _cap_text(value: str, limit: int = _METADATA_TEXT_LIMIT) -> str:
```
Comment (JayantDevkar) [inline api/utils.py:756]: **Ask: add tests.** The suite (1796 tests) doesn't touch `get_tool_summary`'s new fields or `_cap_text` — I verified the behavior manually (exactly-4000 passes untouched; 9000-char input caps with the `[truncated, 9000 chars total]` marker; Edit carries `path`/`old_string`/`new_string`; Task carries `prompt`), but that evidence lives in a review, not the repo. A small parametrized test covering the boundary case and one capped case would lock in the contract the frontend now depends on.

Minor, while you're here: the marker text itself becomes searchable — a timeline search for "truncated" will match every capped event. Probably acceptable; flagging so it's a decision rather than an accident.

File: api/utils.py [hunk 2]
```diff
@@ -780,7 +792,13 @@ def to_relative(path: str) -> str:
         return "Write file", to_relative(path), {"path": path, "content": content}
```
Comment (JayantDevkar) [inline api/utils.py:790]: **Consider capping `content` here too.** `Write.content` still ships uncapped (pre-existing, not introduced by this PR), but it's now the single biggest remaining contributor to the 1s live-poll payload this PR's cap exists to bound — a session that writes a few large files carries all of them in full on every poll. `"content": _cap_text(content)` would close it; the display side already truncates at 2000 chars so nothing visible changes, and search coverage beyond 4000 chars is the same trade-off you already accepted for Edit diffs.

File: frontend/src/lib/utils/timelineLogic.svelte.ts [hunk 3]
```diff
@@ -85,28 +85,92 @@ function matchesSingleFilter(
...
+ * hasHiddenMatch all used to run their own independent walk).
+ *
+ * Invalidated by a cheap content fingerprint, NOT object identity: a live
+ * session's poll (see ConversationView's POLL_INTERVAL_ACTIVE) replaces the
+ * entire event list with freshly-parsed objects every second even when
+ * nothing actually changed, so an identity check would rebuild every
+ * haystack every tick while a search is active. The fingerprint only needs
+ * to change when the *searchable content* changes, not the object reference.
+ */
+const haystackCache = new Map<string, { fingerprint: string; haystack: string }>();
```
Comment (JayantDevkar) [inline frontend/src/lib/utils/timelineLogic.svelte.ts:119]: **Ask: bound this cache.** Module-level `Map`, keyed by `event.id`, never evicted — and the haystack holds the event's *entire* lowercased metadata text (full thinking blocks, results, diffs), so a large session can pin several MB. Karma's main usage pattern is a dashboard tab left open all day navigating across sessions, so this accumulates every session ever searched for the lifetime of the tab.

Cheapest fix given SvelteKit reuses the module across client-side navigations: clear the cache when `createTimelineLogic` is instantiated (new session view → stale entries can never hit again, since ids are session-scoped), or an LRU/size cap if you want cross-navigation reuse. The fingerprint design itself is sound — this is only about lifetime.

File: frontend/src/lib/components/timeline/ToolCallDetail.svelte [hunk 4]
```diff
@@ -240,13 +253,25 @@
 		return val != null && val !== '';
 	}
 
+	// Cap large free-text blocks (Write content, Edit diffs, Task prompts, the
+	// default JSON dump) before display/highlighting — these are only ever
+	// shown in a max-h-80 scroll box anyway, so highlighting the untruncated
+	// string (which can be multi-MB) wastes cycles clipped content never uses.
+	// Copy-to-clipboard buttons still use the original, untruncated string.
+	const DISPLAY_TRUNCATE_LENGTH = 2000;
```
Comment (JayantDevkar) [inline frontend/src/lib/components/timeline/ToolCallDetail.svelte]: **Truncation asymmetry with the server cap.** Server ships up to 4000 chars (`_METADATA_TEXT_LIMIT`) and search matches against all of it, but display clips at 2000 — so a query whose only match lives in chars 2000–4000 shows the amber "match in expanded content" dot, and then expanding shows no highlight anywhere. Same class of gap as the pre-existing 1000-char result display vs full-`result_content` search.

Either align this to 4000 (the strings are already server-bounded, so worst case is the same order of work), or keep 2000 and slice the displayed window around the first match. Tiny issue, but it lands exactly on the "why did this match?" confusion this PR set out to kill.

File: frontend/src/lib/components/timeline/TimelineEventCard.svelte [hunk 5]
```diff
@@ -156,18 +170,26 @@
...
+		// visible (rendered inline via ToolCallDetail or the generic markdown
+		// block below) — the dot would otherwise keep claiming it's still hidden.
+		if (isExpanded) return false;
+		const trimmedQuery = searchQuery.trim();
+		if (!trimmedQuery) return false;
+		const visibleSummary = event.summary
+			? isExpanded
+				? event.summary
+				: truncate(event.summary, 100)
+			: '';
```
Comment (JayantDevkar) [inline frontend/src/lib/components/timeline/TimelineEventCard.svelte]: Nit: dead branch — this whole `$derived` early-returns `false` when `isExpanded` (line 180), so the inner `isExpanded ? event.summary :` ternary can never take the truthy arm. `const visibleSummary = event.summary ? truncate(event.summary, 100) : '';` says the same thing. No behavior change, just removes a head-scratcher.

File: frontend/src/lib/components/timeline/TimelineRail.svelte [hunk 6]
```diff
@@ -204,11 +204,11 @@
...
 
-	// When search query changes, also update TimelineFilterBar's search if searchQuery is set externally
+	// Sync the external (Cmd+F) search query into the timeline's own search state,
+	// including when it's cleared — otherwise Escape closes the search bar but
+	// leaves every card's highlights and the filtered view stuck in place.
 	$effect(() => {
-		if (searchQuery) {
-			timeline.setSearchQuery(searchQuery);
-		}
+		timeline.setSearchQuery(searchQuery);
```
Comment (JayantDevkar) [inline frontend/src/lib/components/timeline/TimelineRail.svelte:210]: Informational, no change needed: the unconditional sync (including clears) is the right call, but note the external path it fixes is currently vestigial — `ConversationView` only ever *clears* `conversationSearchQuery` (Cmd+F toggles `showConversationSearch`, but no input is bound to the query), so this prop is always `''` today. The debounce-cancel effect in `TimelineFilterBar` guards the same not-yet-reachable path. Both are correct-by-construction for when the Cmd+F bar grows an input; flagging so nobody hunts for a search box that doesn't exist yet.

**Review by JayantDevkar — CHANGES_REQUESTED** [review]
In-depth review: read the full diff, traced the data flow server→card→modal→tool panels, and verified the branch hands-on. The core work is solid and every claim in the PR description checked out — requesting changes only for a small set of scoped follow-ups below (tests + robustness), none of which are correctness bugs.

## What was verified

- **API**: `pytest` — 1796 passed; the 1 failure (`test_background_shells_router.py::test_empty_db_returns_empty_list`) is environmental — the suite reads the shared live `~/.claude_karma/metadata.db`, and the "extra" row it found was the reviewer's own test-runner shell. Also unit-exercised `_cap_text`/`get_tool_summary` directly: exactly-4000 string passes untouched, 9000-char `old_string` caps with the `[truncated, 9000 chars total]` marker, `path`/`prompt` fields intact.
- **Frontend**: `npm run check` — 0 errors, 13 warnings, all pre-existing and in files this PR doesn't touch.
- **In-browser** (session `6245d437`, 116 events): search filters + match counter work; the amber hidden-match dot appears exactly on cards whose match is in non-visible content and clears on expand; expanding shows the query highlighted inside the Edit ORIGINAL/REPLACEMENT panels; the previously blank "Editing:" path now renders (the `file_path`/`path` fix).
- **Escape-ordering fix confirmed empirically**: query `& health` (text contains a literal `&`) → 1 match, highlighted, no `&amp;` corruption — this exact query matched 0 events under the old escape-then-match code. Query `(archive` → 4 matches, no regex error (metacharacter escaping works).
- **Live-session cache fix confirmed**: searched a live session while it appended events on the 1s poll — query, highlights, and dot stayed stable as the total climbed 114→116. (A "3 shown vs 1 match" divergence observed during this test is pre-existing tailing behavior — `tailCount = 3` always shows the last 3 events alongside matches — not a PR bug.)

## Asks (all small)

1. Tests for the API metadata change (inline comment on `api/utils.py`).
2. Eviction/bound for the module-level haystack cache (inline comment on `timelineLogic.svelte.ts`).
3. Align display truncation (2000) with the server search cap (4000) or note the gap (inline comment on `ToolCallDetail.svelte`).
4. Consider capping `Write.content` too — it's the biggest remaining uncapped poll-payload contributor and this PR's stated goal (inline comment on `api/utils.py`).

Two further inline notes are informational only (vestigial Cmd+F path, dead ternary) — no action required for approval.

A deeper multi-agent review pass is still running; any additional confirmed findings will arrive as follow-up comments. Happy to re-review as soon as updates land.

🤖 Review assisted by [Claude Code](https://claude.com/claude-code)

File: frontend/src/lib/components/timeline/TimelineEventCard.svelte [hunk 7]
```diff
@@ -156,18 +170,26 @@
...
+		const visibleText = `${displayTitle} ${visibleSummary}`.toLowerCase();
```
Comment (JayantDevkar) [inline frontend/src/lib/components/timeline/TimelineEventCard.svelte]: **Bug: the dot fires on matches that are plainly visible.** `visibleText` counts only `displayTitle` + truncated summary, but the collapsed card also renders the tool-name badge (line ~314), the actor badge (~303), and — for todo events — the todo contents themselves via `TodoUpdateDetail`, *highlighted*, since it now receives `searchQuery`. Meanwhile the haystack includes `tool_name` and `actor`. Net effect, verified by code read: search `bash` → every Bash card shows its "Bash" badge *and* an amber dot claiming the match is hidden; search a subagent's name → same on its actor badge; search todo text on a collapsed ≤3-todo card → the match is visibly highlighted inline while the dot says it's hidden.

Fix direction: include `toolName`, `event.actor`, and (for todo_update events) the todo contents in `visibleText` — or simplest, treat todo_update cards as never-hidden since `TodoUpdateDetail` renders highlights in both collapsed and expanded modes.

File: frontend/src/lib/utils/timelineLogic.svelte.ts [hunk 8]
```diff
@@ -85,28 +85,92 @@ function matchesSingleFilter(
...
+
+function getEventFingerprint(event: TimelineEvent): string {
```
Comment (JayantDevkar) [inline frontend/src/lib/utils/timelineLogic.svelte.ts:124]: **Bug (edge case): fingerprint misses late-arriving searchable metadata → permanently stale haystack.** `spawned_agent_slug` is merged from the `subagent_info` lookup (`api/services/conversation_endpoints.py:361`) and can flip from `null` to a value on a later poll while *every* fingerprinted field stays identical (same timestamp, summary length, result fields). It's not in `SEARCH_SKIP_KEYS`, so it belongs to the haystack — but the stale cache entry keeps serving the pre-slug text, so searching the agent's slug never matches the spawn event for the lifetime of the tab.

Cheap fix that closes the whole class rather than the instance: include a couple of structural signals in the fingerprint — e.g. `Object.keys(m).length` and `(m.spawned_agent_slug ?? '')` — or fingerprint the haystack's own length after a rebuild-on-poll for events whose `has_result` recently flipped. Any of these keeps the 1s-poll cache win while making "metadata grew a field" invalidate correctly.

File: frontend/src/lib/utils/timelineLogic.svelte.ts [hunk 9]
```diff
@@ -85,28 +85,92 @@ function matchesSingleFilter(
...
+
+function getHaystack(event: TimelineEvent): string {
+	const fingerprint = getEventFingerprint(event);
+	const cached = haystackCache.get(event.id);
+	if (cached && cached.fingerprint === fingerprint) {
+		return cached.haystack;
+	}
+	const parts: string[] = [event.title, event.summary ?? '', event.actor];
+	collectSearchableText(event.metadata, parts);
+	const haystack = parts.join(' ').toLowerCase();
```
Comment (JayantDevkar) [inline frontend/src/lib/utils/timelineLogic.svelte.ts]: Nit: joining all fields with a single space lets a multi-word query match across the boundary of two unrelated fields (e.g. `old_string` ending in `foo` + `new_string` starting with `bar` matches the query `foo bar`) — a phantom match the old per-field matcher couldn't produce, and one that shows the hidden-match dot yet highlights nothing when expanded, since highlighting is per-field. Joining with `'\n'` doesn't fix it either — any separator that can appear in a query has this property, but a `'\u0000'` separator (which no realistic query contains) does. One-character change.

File: frontend/src/lib/components/timeline/TimelineEventCard.svelte [hunk 10]
```diff
@@ -60,8 +62,10 @@
...
 
-	// Rendered markdown content for expanded view
-	let renderedExpandedContent = $state('');
+	// Sanitized markdown for expanded view — parsing/sanitizing is the expensive
+	// part, so this effect depends only on content/expansion state, NOT searchQuery.
+	// Highlighting is applied separately below as a cheap $derived over this.
+	let sanitizedExpandedContent = $state('');
 
 	// Render markdown when content changes or expansion state changes
 	$effect(() => {
```
Comment (JayantDevkar) [inline frontend/src/lib/components/timeline/TimelineEventCard.svelte:69]: Efficiency: this effect (and the `highlightHtmlContent` derived over it) runs for expanded **tool_call** events too, but its output is only rendered in the `{:else if}` branch — tool_call expansion always shows `ToolCallDetail` instead. Pre-existing waste, but the new `metadata.prompt` fallback raises the stakes: every expanded Task card now `marked.parse` + `DOMPurify`-sanitizes up to 4KB of prompt nothing displays, and re-walks it per debounced keystroke. Guarding with `event.event_type !== 'tool_call'` in the effect condition (and the derived) skips all of it. Same pattern applies to the popup path in `TimelineRail.svelte`.

File: frontend/src/lib/components/timeline/ToolCallDetail.svelte [hunk 11]
```diff
@@ -457,8 +489,7 @@
 			<div
 				class="font-mono text-sm bg-[var(--bg-muted)] rounded-[var(--radius-md)] p-3 flex items-center justify-between gap-2"
 			>
-				<pre class="text-[var(--event-tool)] overflow-x-auto flex-1">$ {input.command ||
-						''}</pre>
+				<pre class="text-[var(--event-tool)] overflow-x-auto flex-1">$ {#if searchQuery}{@html highlightPlainText(String(input.command || ''), searchQuery)}{:else}{input.command || ''}{/if}</pre>
```
Comment (JayantDevkar) [inline frontend/src/lib/components/timeline/ToolCallDetail.svelte]: Consistency nit: the Bash command block is the only free-text block that skips `truncateForDisplay` — a command embedding a large heredoc gets fully escaped and regex-highlighted on every debounced query change, in a box with no max-height, while the sibling Write/Edit/Task/JSON blocks cap at 2000 chars. Wrapping it like its siblings (`truncateForDisplay(String(input.command || ''))`, keeping the copy button on the full string) closes the exact wasted-cycles case this constant was added to prevent.

File: frontend/src/lib/utils/highlight.ts [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+		if (lastIndex < value.length) {
+			fragment.appendChild(document.createTextNode(value.slice(lastIndex)));
+		}
+		textNode.parentNode?.replaceChild(fragment, textNode);
+	}
+
+	// Re-sanitize: innerHTML was already sanitized once, but reading it back out
+	// and returning it as a fresh HTML string is a parse/serialize round-trip —
+	// re-sanitizing here is the cheap way to stay safe against that class of bug.
+	return DOMPurify.sanitize(container.innerHTML);
```
Comment (JayantDevkar) [inline frontend/src/lib/utils/highlight.ts:95]: Optional: this re-sanitize doubles DOMPurify cost on the per-keystroke hot path this PR was otherwise careful to optimize, and the input is only ever already-sanitized HTML plus `<mark>` nodes built via `createElement`/`textContent` (never `innerHTML`), so there's no reachable injection for it to catch. Returning `container.innerHTML` directly halves the work. That said, defense-in-depth in front of `{@html}` is a defensible trade — your call; if you keep it, consider noting it's belt-and-suspenders so a future optimizer doesn't puzzle over it.

**Review by JayantDevkar — COMMENTED** [review]
Follow-up to my earlier review — the deeper multi-agent pass finished and surfaced additional findings; each below was independently re-verified against the branch before posting. Two are real behavior bugs in the new feature (the hidden-match dot false-positive and the fingerprint staleness window); the rest are efficiency/consistency nits. The earlier CHANGES_REQUESTED verdict stands as the gate — treat the two bugs as part of that bar, the nits as judgment calls.

One finding has no diff line to anchor on: **`TimelineRail.svelte` ends with a stray markdown code fence** — the last line of the file is a literal ``` ``` ``` after the final `{/if}`, which Svelte renders as a text node at the bottom of every timeline. It's pre-existing on `main` (not caused by this PR), but this PR touches the file heavily, so sweeping it here is one deleted line.

Also a general style note rather than an inline thread: the multi-paragraph rationale comments added across this PR (the 14-line haystack-cache block in `timelineLogic.svelte.ts:124-137` is the largest; similar blocks in `TimelineEventCard`, `TimelineFilterBar`, `TimelineRail`, `ToolCallDetail`, `api/utils.py`) are written for the *reviewer of this diff* — cache-invalidation war stories, debounce race narratives. This repo keeps comments to short single-line pointers; rationale that justifies the change belongs in the PR description (where it already is, well written). Trimming each block to one line would match the surrounding code.

🤖 Review assisted by [Claude Code](https://claude.com/claude-code)

File: api/utils.py (hunk 1 above)
Comment (the-non-expert) [reply api/utils.py:756]: Added `api/tests/test_utils_tool_summary.py` (10 tests) covering the boundary case (exactly-4000 passes untouched), the capped case (9000 chars → truncation marker), and that Edit/Write/Task carry their capped fields correctly. Left the marker-text-is-searchable behavior as-is per your framing — a decision, not an accident. Fixed in 474f36a.

File: api/utils.py (hunk 2 above)
Comment (the-non-expert) [reply api/utils.py:790]: Done — `content = _cap_text(tool_input.get("content", ""))` in the Write branch. Fixed in 474f36a.

File: frontend/src/lib/utils/timelineLogic.svelte.ts (hunk 3 above)
Comment (the-non-expert) [reply frontend/src/lib/utils/timelineLogic.svelte.ts:119]: Went with the cheap option — `haystackCache.clear()` at the top of `createTimelineLogic`, since it's only ever instantiated once per session view and ids are session-scoped so stale entries could never hit anyway. Fixed in 474f36a.

File: frontend/src/lib/components/timeline/ToolCallDetail.svelte (hunk 4 above)
Comment (the-non-expert) [reply frontend/src/lib/components/timeline/ToolCallDetail.svelte]: Aligned `DISPLAY_TRUNCATE_LENGTH` to 4000 to match the server's `_METADATA_TEXT_LIMIT`. Fixed in 474f36a.

File: frontend/src/lib/components/timeline/TimelineEventCard.svelte (hunk 5 above)
Comment (the-non-expert) [reply frontend/src/lib/components/timeline/TimelineEventCard.svelte]: Removed — folded into the `hasHiddenMatch` rewrite for the false-positive bug below, which no longer has the dead ternary. Fixed in 474f36a.

File: frontend/src/lib/components/timeline/TimelineRail.svelte (hunk 6 above)
Comment (the-non-expert) [reply frontend/src/lib/components/timeline/TimelineRail.svelte:210]: Acknowledged, no code change needed — resolving since you flagged this as informational only.

File: frontend/src/lib/components/timeline/TimelineEventCard.svelte (hunk 7 above)
Comment (the-non-expert) [reply frontend/src/lib/components/timeline/TimelineEventCard.svelte]: Fixed — `hasHiddenMatch` now builds `visibleText` from title, summary (skipped for todo_update, which doesn't render it), toolName (when the badge shows), actor (when the badge shows), and todo contents when ≤3 todos render inline (mirroring `TodoUpdateDetail`'s own `inlineLimit`). Fixed in 474f36a.

File: frontend/src/lib/utils/timelineLogic.svelte.ts (hunk 8 above)
Comment (the-non-expert) [reply frontend/src/lib/utils/timelineLogic.svelte.ts:124]: Went with your first suggestion — added `Object.keys(m).length` and `m.spawned_agent_slug ?? ''` to the fingerprint. Fixed in 474f36a.

File: frontend/src/lib/utils/timelineLogic.svelte.ts (hunk 9 above)
Comment (the-non-expert) [reply frontend/src/lib/utils/timelineLogic.svelte.ts]: Switched the join separator to `String.fromCharCode(0)` (NUL) — one-character change as you noted. Fixed in 474f36a.

File: frontend/src/lib/components/timeline/TimelineEventCard.svelte (hunk 10 above)
Comment (the-non-expert) [reply frontend/src/lib/components/timeline/TimelineEventCard.svelte:69]: Guarded both the effect and the derived with `event.event_type !== 'tool_call' && event.event_type !== 'todo_update'` in both `TimelineEventCard.svelte` and the popup path in `TimelineRail.svelte`. Fixed in 474f36a.

File: frontend/src/lib/components/timeline/ToolCallDetail.svelte (hunk 11 above)
Comment (the-non-expert) [reply frontend/src/lib/components/timeline/ToolCallDetail.svelte]: Wrapped it — `truncateForDisplay(String(input.command || ''))`, copy button still uses the untruncated `input.command`. Fixed in 474f36a.

File: frontend/src/lib/utils/highlight.ts (hunk 12 above)
Comment (the-non-expert) [reply frontend/src/lib/utils/highlight.ts:95]: Kept the re-sanitize — didn't want to trade away defense-in-depth on an `{@html}` path for a minor perf win. Trimmed the comment to make clear it's an intentional belt-and-suspenders call, not leftover work. Fixed in 474f36a.

**Review by JayantDevkar — APPROVED** [review]
Re-verified at `474f36a` — every ask from both review rounds is addressed, and each fix was confirmed against the actual code and a live run, not just the replies. Approving.

## Verification of the fixes

- **API tests**: `tests/test_utils_tool_summary.py` — 10/10 pass, covering exactly the contract that matters (4000-boundary untouched, over-limit gets the marker, Edit/Write/Task carry capped fields). Full suite: **1806 passed** (up 10), same single pre-existing environmental failure (`test_background_shells_router` reads the shared live DB and found this reviewer's own shell — unrelated to the branch).
- **`Write.content` capped** — confirmed in `get_tool_summary`.
- **Haystack cache**: `haystackCache.clear()` on `createTimelineLogic` instantiation — correct choice given session-scoped event ids; fingerprint now includes `Object.keys(m).length` + `spawned_agent_slug`, closing the late-arriving-metadata staleness class, not just the instance.
- **Hidden-match dot**: `visibleText` now mirrors what the collapsed card actually renders (tool badge, actor badge, ≤3 inline todos matching `TodoUpdateDetail`'s `inlineLimit`). Verified in-browser: searching `bash` over a 116-event session → 13 matches, zero false dots on Bash-badged cards (previously every one showed the dot).
- **Truncation aligned** to 4000 = `_METADATA_TEXT_LIMIT`; Bash commands now truncate like their sibling blocks with copy preserving the full string.
- **Cross-field phantom**: NUL separator — verified one-char change.
- **Markdown pipeline** guarded for `tool_call`/`todo_update` in both the card effect/derived and the rail popup path.
- **Stray fence** removed from `TimelineRail.svelte`; comment blocks trimmed throughout; the kept re-sanitize in `highlight.ts` is clearly labeled intentional — reasonable call.
- `npm run check`: **0 errors**, same 13 pre-existing warnings in untouched files.

Nice turnaround — the fixes consistently addressed the *class* of each issue rather than the instance.

🤖 Review assisted by [Claude Code](https://claude.com/claude-code)
